common/argparse.cpp

```diff
@@ -0,0 +1,1362 @@
+  }
+  return true;
+}
+
+auto Args::Parser::ParseStringArgValue(const Arg& arg, llvm::StringRef value)
+    -> bool {
+  CARBON_CHECK(arg.kind == Arg::Kind::String) << "Incorrect kind: " << arg.kind;
+  if (!arg.is_append) {
+    *arg.string_storage = value;
+  } else {
+    arg.string_sequence->push_back(value);
+  }
+  return true;
+}
+
+auto Args::Parser::ParseOneOfArgValue(const Arg& arg, llvm::StringRef value)
+    -> bool {
+  CARBON_CHECK(arg.kind == Arg::Kind::OneOf) << "Incorrect kind: " << arg.kind;
+  if (!arg.value_action(arg, value)) {
+    std::string v;
+    llvm::raw_string_ostream vs(v);
+    llvm::printEscapedString(value, vs);
+    errors_ << "ERROR: Option '--" << arg.info.name << "=" << v
+            << "' has an invalid value '" << v << "'; valid values are: ";
+    for (auto value_string : arg.value_strings.drop_back()) {
+      errors_ << "'" << value_string << "', ";
+    }
+    if (arg.value_strings.size() > 1) {
+      errors_ << "or ";
+    }
+    errors_ << "'" << arg.value_strings.back() << "'\n";
+    return false;
+  }
+  return true;
+}
+
+auto Args::Parser::ParseArg(const Arg& arg,
+                            std::optional<llvm::StringRef> value,
+                            bool negated_name) -> bool {
+  // If this argument has a meta action, replace the current meta action with
+  // it.
+  if (arg.meta_action) {
+    arg_meta_action_ = arg.meta_action;
+  }
+
+  // Boolean flags have special parsing logic.
+  if (negated_name) {
+    return ParseNegatedFlag(arg, value);
+  }
+  if (arg.kind == Arg::Kind::Flag) {
+    return ParseFlag(arg, value);
+  }
+
+  if (!value) {
+    // We can't have a positional argument without a value, so we know this is
+    // an option and handle it as such.
+    if (arg.kind == Arg::Kind::MetaActionOnly) {
+      // Nothing further to do here, this is only a meta-action.
+      return true;
+    }
+    if (!arg.has_default) {
+      errors_ << "ERROR: Option '--" << arg.info.name
+              << "' requires a value to be provided and none was.\n";
+      return false;
+    }
+    SetOptionDefault(arg);
+    return true;
+  }
+
+  // There is a value to parse as part of the argument.
+  switch (arg.kind) {
+    case Arg::Kind::Integer:
+      return ParseIntegerArgValue(arg, *value);
+    case Arg::Kind::String:
+      return ParseStringArgValue(arg, *value);
+    case Arg::Kind::OneOf:
+      return ParseOneOfArgValue(arg, *value);
+    case Arg::Kind::MetaActionOnly:
+      errors_ << "ERROR: Option '--" << arg.info.name
+              << "' cannot be used with a value, and '" << *value
+              << "' was provided.\n";
+      // TODO: improve message
+      return false;
+    case Arg::Kind::Flag:
+    case Arg::Kind::Invalid:
+      CARBON_FATAL() << "Invalid kind!";
+  }
+}
+
+auto Args::Parser::SplitValue(llvm::StringRef& unparsed_arg)
+    -> std::optional<llvm::StringRef> {
+  // Split out a value if present.
+  std::optional<llvm::StringRef> value = {};
```

Comment:
Why the explicit initialization on things like this?

Reply:
(I thought I had changed all of these, but now I have.)